pkg/controllers/mcs/mcs_controller.go

```diff
@@ -0,0 +1,380 @@
+
+// CrossClusterEventReason is indicates the reason of CrossCluster event.
+const CrossClusterEventReason string = "MCSCrossCluster"
+
+// MCSServiceAppliedConditionType is indicates the condition type of mcs service applied.
+const MCSServiceAppliedConditionType = "ServiceApplied"
+
+// MCSWorkIDLabel is indicates the label of work created by mcs.
+const MCSWorkIDLabel = "mcs.karmada.io/mcs-id"
+
+// MCSController is to sync MultiClusterService.
+type MCSController struct {
+	client.Client
+	EventRecorder      record.EventRecorder
+	RateLimiterOptions ratelimiterflag.Options
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+// The Controller will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (c *MCSController) Reconcile(
+	ctx context.Context,
+	req controllerruntime.Request,
+) (controllerruntime.Result, error) {
+	klog.V(4).InfoS("Reconciling MultiClusterService", "namespace", req.Namespace, "name", req.Name)
+
+	mcs := &networkingv1alpha1.MultiClusterService{}
+	if err := c.Client.Get(ctx, req.NamespacedName, mcs); err != nil {
+		if apierrors.IsNotFound(err) {
+			// The mcs no longer exist, in which case we stop processing.
+			return controllerruntime.Result{}, nil
+		}
+		klog.ErrorS(err, "failed to get MultiClusterService object", "NamespacedName", req.NamespacedName)
+		return controllerruntime.Result{}, err
+	}
+
+	if !mcs.DeletionTimestamp.IsZero() {
+		return c.handleMCSDelete(ctx, mcs.DeepCopy())
+	}
+	return c.handleMCSCreateOrUpdate(ctx, mcs.DeepCopy())
+}
+
+func (c *MCSController) handleMCSDelete(
+	ctx context.Context,
+	mcs *networkingv1alpha1.MultiClusterService,
+) (controllerruntime.Result, error) {
+	klog.V(4).InfoS("Begin to handle MultiClusterService delete event", "namespace", mcs.Namespace, "name", mcs.Name)
+
+	if _, err := c.deleteServiceWork(ctx, mcs, sets.New[string]()); err != nil {
+		c.EventRecorder.Event(mcs, corev1.EventTypeWarning, CrossClusterEventReason,
+			fmt.Sprintf("failed to delete service work :%v", err))
+		return controllerruntime.Result{}, err
+	}
+
+	finalizersUpdated := controllerutil.RemoveFinalizer(mcs, MCSControllerFinalizer)
+	if finalizersUpdated {
+		err := c.Client.Update(ctx, mcs)
+		if err != nil {
+			klog.V(4).ErrorS(err, "failed to update MultiClusterService with finalizer",
+				"namespace", mcs.Namespace, "name", mcs.Name)
+			return controllerruntime.Result{}, err
+		}
+	}
+
+	klog.V(4).InfoS("Success to delete MultiClusterService", "namespace", mcs.Namespace, "name", mcs.Name)
+	return controllerruntime.Result{}, nil
+}
+
+func (c *MCSController) deleteServiceWork(
+	ctx context.Context,
+	mcs *networkingv1alpha1.MultiClusterService,
+	retainClusters sets.Set[string],
+) (controllerruntime.Result, error) {
+	workList, err := helper.GetWorksByLabelsSet(c, labels.Set{MCSWorkIDLabel: string(mcs.UID)})
```

Comment:
Let's generate a id for mcs, referring to: https://github.com/karmada-io/karmada/blob/1c10def029d9f3409f4530352423f48a193bac0d/pkg/detector/detector.go#L667